inc/cappuccino/TSet.hpp

```diff
@@ -0,0 +1,520 @@
+     * @return The number of elements inserted based on `allow`.
+     */
+    template <typename RangeType>
+    auto InsertRange(
+        RangeType&& key_range,
+        std::chrono::milliseconds ttl,
+        Allow allow = Allow::INSERT_OR_UPDATE) -> size_t;
+
+    /**
+     * Inserts or updates a range of keys with default TTL. 
+     * @tparam RangeType A container of KeyTypes.
+     * @param key_range The elements to insert or update into the set.
+     * @param allow Allowed methods of insertion | update.  Defaults to allowing
+     *              insertions and updates.
+     * @return The number of elements inserted based on `allow`.
+     */
+    template <typename RangeType>
+    auto InsertRange(
+        RangeType&& key_range,
+        Allow allow = Allow::INSERT_OR_UPDATE) -> size_t;
+
+    /**
+     * Attempts to delete the given key.
+     * @param key The key to remove from the set.
+     * @return True if the key was deleted, false if the key does not exist.
+     */
+    auto Delete(
+        const KeyType& key) -> bool;
+
+    /**
+     * Attempts to delete all given keys.
+     * @tparam RangeType A container with the set of keys  to delete, e.g. vector<k> or set<k>.
+     * @param key_range The keys to delete from the set.
+     * @return The number of items deleted from the set.
+     */
+    template <template <class...> typename RangeType>
+    auto DeleteRange(
+        const RangeType<KeyType>& key_range) -> size_t;
+
+    /**
+     * Attempts to find the given key.
+     * @param key The key to lookup.
+     * @return True if key exists, or false if it does not.
+     */
+    auto Find(const KeyType& key) -> bool;
+
+    /**
+     * Attempts to find all the given keys presence.
+     * @tparam RangeType A container with the set of keys to lookup, e.g. vector<KeyType>.
+     * @param key_range A container with the set of keys to lookup.
+     * @return All input keys with a bool indicating if it exists.
+     */
+    template <typename RangeType>
+    auto FindRange(const RangeType& key_range) -> std::vector<std::pair<KeyType, bool>>;
+
+    /**
+     * Attempts to find all given keys presence.
+     *
+     * The user should initialize this container with the keys to lookup with the values all
+     * bools. The keys that are found will have the bools set indicating presence in the set.
+     *
+     * @tparam RangeType A container with a pair of optional items,
+     *                   e.g. vector<pair<k, bool>> or map<k, bool>.
+     * @param key_bool_range The keys to bools to fill out.
+     */
+    template <typename RangeType>
+    auto FindRangeFill(
+        RangeType& key_bool_range) -> void;
+
+    /**
+     * Trims the TTL list of items and evicts all expired elements.
+     * @return The number of elements pruned.
+     */
+    auto CleanExpiredValues() -> size_t;
+
+    /**
+     * @return If this set is currently empty.
+     */
+    auto empty() const -> bool { return m_used_size == 0; }
```

Comment:
The more I look at m_used_size the more I think it should be std::atomic (for _every_ cache in this library) and that its currently kind of a bug and works mostly because there are a lot of mutexes in use.  This read though should ideally be an std::atomic .load(std::memory_order_relaxed) to guarantee its the correct value in all use cases.  If we do end up switching this to std::atomic<size_t> then every ++ or -- operation should be switched to fetch_add or fetch_sub with std::memory_order_relaxed

Reply:
I agree I realized this was a bit questionable when I wanted to also have size() and empty() evict from the cache but then that called constness into question. At that point I realized the m_used_size has potential to just be completely stale.

Reply:
Can you get rid of m_used_size entirely since you can just use the map/set for the actual number of items in the map or set?

Reply:
Yes, good idea!